add dependencies for cryptsetup

bcressey opened the pull request:
**Issue number:**
Related: #327

**Description of changes:**
Add dependencies for cryptsetup. The integration with systemd is postponed until #438 is merged.

I'd intended to include only the cryptsetup libraries, but it was difficult to test the functionality without the actual binaries (`cryptsetup`, `integritysetup`, `veritysetup`). Likewise for libtss2 - the various `tpm2_*` tools are useful to prove the libraries still work.

These changes are limited to adding new packages so that the pull request will be safe to merge without needing to scrutinize any changes on the systemd side.

**Testing done:**
`cryptsetup` and `systemd-cryptenroll`  work as expected:
```
# cryptsetup luksFormat --pbkdf argon2id volume 

WARNING!
========
This will overwrite data on volume irrevocably.

Are you sure? (Type 'yes' in capital letters): YES
Enter passphrase for volume: 
Verify passphrase: 

# systemd-cryptenroll --tpm2-device=auto volume

🔐 Please enter current passphrase for disk volume: ****                    
New TPM2 token enrolled as key slot 1.

# systemd-cryptenroll volume 
SLOT TYPE    
   0 password
   1 tpm2

# /usr/lib/systemd/systemd-cryptsetup attach crypt volume none tpm2-device=auto

# ls -latr /dev/mapper/crypt 
brw-rw----. 1 root root 252, 1 Mar 23 21:03 /dev/mapper/crypt
```

Various `tpm2` tools also work:
```
bash-5.1# tpm2_pcrread sha256:7+sha512:7
  sha256:
    7 : 0xD5A213BBD7D72E5EEADBE8C4DC13078B6D63A5B372FD33CA9BF17C1BC8AE89DB
  sha512:
    7 : 0xF15855D78F2984690339540A809EAE43DF9F5642611340175AFCD226991F35CF1F08A119061A2B7B01D3BCE6628D9CD99BB0FDF5419C5D0030F4C250D1F0926F
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

mikn committed:
packages: add libcrypto

Co-authored-by: Mikael Knutsson (mikn) <mikael.knutsson@gmail.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>

mikn committed:
packages: add cryptsetup

Co-authored-by: Mikael Knutsson (mikn) <mikael.knutsson@gmail.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>

mikn committed:
packages: add libdevmapper

Co-authored-by: Mikael Knutsson (mikn) <mikael.knutsson@gmail.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>

mikn committed:
packages: add libaio

Co-authored-by: Mikael Knutsson (mikn) <mikael.knutsson@gmail.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>

mikn committed:
packages: add libpopt

Co-authored-by: Mikael Knutsson (mikn) <mikael.knutsson@gmail.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>

mikn committed:
packages: add libtss2

Co-authored-by: Mikael Knutsson (mikn) <mikael.knutsson@gmail.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>

Commit:
packages: add tpm2-tools

Signed-off-by: Ben Cressey <bcressey@amazon.com>

arnaldo2792 commented on packages/tpm2-tools/tpm2-tools.spec:
```diff
@@ -0,0 +1,152 @@
+Patch0001: 0001-tpm2-tools-disable-SM2-and-SM3-checks.patch
+
+# libcurl isn't available
+Patch0002: 0002-tpm2-tools-disable-tpm2_getekcertificate.patch
```
nit: the commit message in the patch could explain that `tpm2_getekcertificate` isn't needed due to the lack of `libcurl`.

arnaldo2792 commented on packages/libpopt/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[[package.metadata.build-package.external-files]]
+url = "http://ftp.rpm.org/popt/releases/popt-1.x/popt-1.19.tar.gz"
+sha512 = "5d1b6a15337e4cd5991817c1957f97fc4ed98659870017c08f26f754e34add31d639d55ee77ca31f29bb631c0b53368c1893bd96cf76422d257f7997a11f6466"
+force-upstream = true
```
Is this required? Is there a problem with us distributing these sources?

bcressey replied:
> Is this required? Is there a problem with us distributing these sources?

Oops, I missed this - I'll remove it in the next revision.

arnaldo2792 commented on packages/cryptsetup/cryptsetup.spec:
```diff
@@ -0,0 +1,89 @@
+%build
+%cross_configure \
+  --disable-asciidoc \
+  --disable-fips \
```
Does this mean that there will be a separate copy of `cryptsetup` that enables FIPS?

bcressey replied:
I'm not sure the best way to handle this.

For AWS-LC, I'm not confident that builds with and without FIPS enabled at compile time will always be ABI-compatible.

For cryptsetup, I could build the binaries with and without FIPS, but I'm reluctant to do that for the libraries, and without that the FIPS checks won't necessarily be enforced.

arnaldo2792 replied:
Oh FIPS :melting_face:

bcressey commented:
⬆️ force push to rebase (no changes to these commits)

bcressey commented:
⬆️ force push addresses feedback from @arnaldo2792

On `aws-k8s-1.32-fips` with the `cryptsetup` package added, the test case now prints:
```
# cryptsetup luksFormat --pbkdf argon2id volume 

WARNING!
========
This will overwrite data on volume irrevocably.

Are you sure? (Type 'yes' in capital letters): YES
Enter passphrase for volume: 
Verify passphrase: 
Only PBKDF2 is supported in FIPS mode.
Failed to set pbkdf parameters.
```

On a non-FIPS variant, argon2 can still be used.

arnaldo2792 commented on packages/libcrypto/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+path = "../packages.rs"
+
+[package.metadata.build-package]
+releases-url = "https://github.com/aws/aws-lc/commits/fips-2024-09-27/"
```
I wonder if we will be able to vend the one version of AWS-LC that all Bottlerocket components consume, with this now we have 3 (aws-lc-rs, libcrypto in the SDK for `gofips`). Doesn't harm but it might be confusing.

arnaldo2792 commented on packages/libtss2/tss-sysusers.conf:
```diff
@@ -0,0 +1,2 @@
+# User for TPM devices
+u tss 59
```
Was this user selected to match what other Linux distros use?

bcressey replied:
> Was this user selected to match what other Linux distros use?

Yes, I picked the same one as Fedora.

KCSesh commented on packages/tpm2-tools/0001-tpm2-tools-disable-SM2-and-SM3-checks.patch:
```diff
@@ -0,0 +1,51 @@
+From 7c0afbd0f871c1973876dc1c20648c82d01ad84d Mon Sep 17 00:00:00 2001
+From: Ben Cressey <bcressey@amazon.com>
+Date: Sat, 22 Mar 2025 16:57:25 +0000
+Subject: [PATCH 1/2] tpm2-tools: disable SM2 and SM3 checks
```
```suggestion
Subject: [PATCH] tpm2-tools: disable SM2 and SM3 checks
```

KCSesh commented on packages/tpm2-tools/0002-tpm2-tools-disable-tpm2_getekcertificate.patch:
```diff
@@ -0,0 +1,54 @@
+From ea01f54f0a108e35149f9b17e622dd20cc549e92 Mon Sep 17 00:00:00 2001
+From: Ben Cressey <bcressey@amazon.com>
+Date: Sat, 22 Mar 2025 16:59:46 +0000
+Subject: [PATCH 2/2] tpm2-tools: disable tpm2_getekcertificate
```
```suggestion
Subject: [PATCH] tpm2-tools: disable tpm2_getekcertificate
```

ytsssun commented on packages/cryptsetup/0001-pbkdf-check-whether-FIPS-is-enabled-at-runtime.patch:
```diff
@@ -0,0 +1,55 @@
++	f = fopen("/proc/sys/crypto/fips_enabled", "r");
++	if (f) {
++		r = fscanf(f, "%1d", &fips_enabled);
++		if (r != 1)
++			fips_enabled = 0;
```
Is it expected to default to non-fips mode if file open failed. Also would it make sense to log the file open behavior instead of silently fallback to non-fips mode?

bcressey replied:
> Is it expected to default to non-fips mode if file open failed. Also would it make sense to log the file open behavior instead of silently fallback to non-fips mode?

If a kernel doesn't enable FIPS at all, then that file might not exist (and the kernel is not in FIPS mode).

I don't really want to add more logging here, since it'll change the output of what should be a standard program in ways that are unique to Bottlerocket.

bcressey commented:
⬆️ force push fixes patch numbering per @KCSesh
